core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -88,7 +91,18 @@ private ParallelIterator(
     @Override
     public void close() {
       // close first, avoid new task submit
-      this.closed = true;
+      this.closed.set(true);
+
+      for (Task<T> task : yieldedTasks) {
+        try {
+          task.close();
+        } catch (Exception e) {
+          throw new RuntimeException("Close failed", e);
+        }
+      }
+      yieldedTasks.clear();
+
+      // TODO close input iterables that were not started yet
```

Comment:
This is preexisting problem and easy to fix. Let's fix as a follow-up to keep the PRs easy to review.